### PR TITLE
new module basic adjustments

### PR DESCRIPTION
**Pull request by edgardoh**
This new module allows to set the exposure, highlight recovery, contrast, brightness and saturation.
It has an auto exposure option that works on the entire image or a user defined area.

Is an alternative to the base curve, usually I get nicer results for a edit starting point.

The auto option and highlight recovery are based on RawTherapee auto levels & highlight recovery.

### CODE REVIEW

**Comment by TurboGit** [comment]
Sounds good! I'll test this one.

At some point I feel that we should have a "beginner" module group and this IOP should be part of this group.

**Comment by TurboGit** [comment]
Maybe use a % for sliders contrast, brightness and saturation.

For exposure 4 digits are too much, 2 will be certainly enough (as in exposure module). This will make this simple module even more beginner friendly.

If you agree and don't have time I can handle this?

**Comment by edgardoh** [comment]
I'm ok with 2 digits for exposure, but I'm not sure about the %, the
current contrast, brightness saturation don't use a %, shouldn't this be
the same?

**Comment by edgardoh** [comment]
I just moved all to 2 digits, seems nice this way.

**Comment by aurelienpierre** [comment]
Sorry, I don't like that.

Exposure adjustments should come before the input color profile, but the color adjustments should come after. So you can't bundle them together in the same IOP.

Then, the lightness adjustment is a simple gamma on decorrelated RGB channels, which is bad especially this early in the pipe.

Also, you are duplicating the fulcrum contrast and saturation from the color balance. I planned on adding a vibrance and chroma adjustments, maybe we could keep it the color hub of dt, as Capture One or Da Vinci Resolve do ?

On another note, when you write:

```C 
  // exposure
  pixel.x = (pixel.x - black_point) * scale;
  pixel.y = (pixel.y - black_point) * scale;
  pixel.z = (pixel.z - black_point) * scale;
```

you should write full vectors operations, like:

```C
pixel = (pixel - black_point) * scale;
```

because OpenCL is designed to be vectorized (and I don't know if the compiler is clever enough to detect you are working on vector coordinates and pack them).

**Comment by edgardoh** [comment]
@TurboGit , this has the wip label but I'm not sure what's expected to change here.

**Comment by TurboGit** [comment]
Just that I did not visit this PR since some time. Not wip, so I'll remove the flag. BTW, this needs a conflict resolution.

**Comment by TurboGit** [comment]
@edgardoh : also Aurélien's comment about pixel vectorization does not seem to have been taken into account. Any issue with that?

**Comment by edgardoh** [comment]
Thanks, I did not mean to rush you, just an FYI.

I'll upgrade this.

**Comment by edgardoh** [comment]
@TurboGit , all fixed.

File: src/iop/basicadj.c [hunk 1]
```diff
@@ -0,0 +1,1744 @@
...
+    // gavg += histogram[ii] * _get_LUTf(gamma2curve, gamma2curve_size, val);
+    gavg += histogram[ii] * gamma2(val);
+    val += increment;
+  }
+
+  gavg /= sum;
+
+  if(black < gavg)
+  {
+    int maxwhiteclip = (gavg - black) * 4 / 3
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const

File: src/iop/basicadj.c [hunk 2]
```diff
@@ -0,0 +1,1744 @@
...
+  // compute contrast that spreads the average spacing of octiles
+  contr = 50.0f * (1.1f - ospread);
+  contr = MAX(0.f, MIN(100.f, contr));
+  // take gamma into account
+  double whiteclipg = gamma2(whiteclip * corr);
+
+  float gavg = 0.f;
+
+  float val = 0.f;
+  float increment = corr * (1 << histcompr);
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const

File: src/iop/basicadj.c [hunk 3]
```diff
@@ -0,0 +1,1744 @@
...
+
+#include "bauhaus/bauhaus.h"
+#include "develop/imageop.h"
+#include "gui/color_picker_proxy.h"
+
+DT_MODULE_INTROSPECTION(1, dt_iop_basicadj_params_t)
+
+#define exposure2white(x) exp2f(-(x))
+
+typedef enum dt_iop_basicadj_preservecolors_t {
```
Comment (TurboGit) [inline src/iop/basicadj.c]: { should be on new line

File: src/iop/basicadj.c [hunk 4]
```diff
@@ -0,0 +1,1744 @@
...
+    double f;
+    int64_t i;
+  } tmp;
+  tmp.i = i;
+  return tmp.f;
+}
+
+static inline int ilogbp1(double d)
+{
+  int m = d < 4.9090934652977266E-91;
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const

File: src/iop/basicadj.c [hunk 5]
```diff
@@ -0,0 +1,1744 @@
...
+  u = longBitsToDouble(((int64_t)(q + 0x3ff)) << 52);
+  return x * u;
+}
+
+static inline double xlog(double d)
+{
+  double x, x2, t, m;
+  int e;
+
+  e = ilogbp1(d * 0.7071);
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const int e = ...

File: src/iop/basicadj.c [hunk 6]
```diff
@@ -0,0 +1,1744 @@
...
+  m = ldexpk(d, -e);
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const double m = ...

File: src/iop/basicadj.c [hunk 7]
```diff
@@ -0,0 +1,1744 @@
...
+  float expcomp = 0.f;
+  float black = 0.f;
+  float bright = 0.f;
+  float contr = 0.f;
+  float hlcompr = 0.f;
+  float hlcomprthresh = 0.f;
+
+  float scale = 65536.0f;
+
+  int imax = 65536 >> histcompr;
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const

File: src/iop/basicadj.c [hunk 8]
```diff
@@ -0,0 +1,1744 @@
...
+  rawmax <<= histcompr;
+  whiteclip <<= histcompr;
+  ave = ave * (1 << histcompr);
+  median <<= histcompr;
+  shc <<= histcompr;
+
+  // compute exposure compensation as geometric mean of the amount that
+  // sets the mean or median at middle gray, and the amount that sets the estimated top
+  // of the histogram at or near clipping.
+  float expcomp1 = (log(midgray * scale / (ave - shc + midgray * shc))) / log(2.f);
```
Comment (TurboGit) [inline src/iop/basicadj.c]: const ?

File: src/iop/basicadj.c (hunk 3 above)
Comment (edgardoh) [reply src/iop/basicadj.c]: We still have this clang issue, even if I change it, clang will change it back.

**Comment by edgardoh** [comment]
Review changes ready.

**Comment by TurboGit** [comment]
Thanks! Looks good merged.

**Comment by PaoloAst** [comment]
Great job! I don't see any effect of "middle grey" slider...

**Comment by edgardoh** [comment]
Thanks.

The middle grey works with the contrast slider, it defines the lightness
that is not modified by the contrast. The default is the middle grey for
the work rgb colorspace. If the work profile changes, the default will
change too, but you'll need to reset it by hand.

**Comment by PaoloAst** [comment]
Thanks @edgardoh !

**Comment by GLLM** [comment]
> This new module allows to set the exposure, highlight recovery, contrast, brightness and saturation.
> It has an auto exposure option that works on the entire image or a user defined area.
> 
> Is an alternative to the base curve, usually I get nicer results for a edit starting point.
> 
> The auto option and highlight recovery are based on RawTherapee auto levels & highlight recovery.

Hello @edgardoh 
Can I ask (couldnt find it in the doc) what is the difference between the Exposure and the Brightness effects ?
Moreover, the Middle grey slider does not do anything ? What is it meant for ? What can I possibly do wrong ?
Thanks a lot.
GLLM

**Pull request merged**